RavenDB-14882 : Enhance time-series client API

aviv86 opened the pull request:
https://issues.hibernatingrhinos.com/issue/RavenDB-14882
- Have separate command for getting multi time-series and only one time-series
- Get non-existing TS should return null instead an empty array (both from session and from GetTimeSeriesOperation)
- Evict TS upon entity eviction

aviv86 committed:
RavenDB-14882 : Have separate command for getting multi time-series and only one time-series

aviv86 committed:
RavenDB-14882 : Get non-existing TS should return null instead an empty array

aviv86 committed:
RavenDB-14882 : apply changes to tests

aviv86 committed:
RavenDB-14882 : Evict TS upon entity eviction

aviv86 committed:
RavenDB-14882 : fix failing tests

karmeli87 commented on src/Raven.Client/Documents/Operations/TimeSeries/GetMultipleTimeSeriesOperation.cs:
```diff
@@ -0,0 +1,122 @@
+                    pathBuilder.Append("&name=")
+                        .Append(range.Name)
+                        .Append("&from=")
+                        .Append(range.From.GetDefaultRavenFormat())
```
We need to make sure this in UTC.
(We can do that after https://github.com/ravendb/ravendb/pull/10517 is merged)

karmeli87 commented on src/Raven.Client/Documents/Operations/TimeSeries/GetTimeSeriesOperation.cs:
```diff
@@ -88,15 +70,12 @@ public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, Serve
+                pathBuilder.Append("&name=")
+                    .Append(_name)
+                    .Append("&from=")
+                    .Append(_from.GetDefaultRavenFormat())
```
Same

aviv86 replied:
done

karmeli87 commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -1251,6 +1251,7 @@ public void Evict<T>(T entity)
                 DocumentsByEntity.Evict(entity);
                 DocumentsById.Remove(documentInfo.Id);
                 _countersByDocId?.Remove(documentInfo.Id);
+                _timeSeriesByDocId?.Remove(documentInfo.Id);
```
Do we need to clear the cache as well?

aviv86 replied:
what do you mean? this is the time series cache

karmeli87 replied:
Right, somehow I thought there a different entity for TS caching.

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/GetMultipleTimeSeriesOperation.cs:
```diff
@@ -0,0 +1,122 @@
+                        .Append(_pageSize);
+                }
+
+                foreach (var range in _ranges)
```
Can this be empty?

aviv86 replied:
we throw an exception if this is empty

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/GetMultipleTimeSeriesOperation.cs:
```diff
@@ -0,0 +1,122 @@
+                foreach (var range in _ranges)
+                {
+                    pathBuilder.Append("&name=")
+                        .Append(range.Name)
```
we need to escape this

aviv86 replied:
done

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/GetTimeSeriesOperation.cs:
```diff
@@ -88,15 +70,12 @@ public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, Serve
-                        .Append(range.To.GetDefaultRavenFormat());
-                }
+                pathBuilder.Append("&name=")
+                    .Append(_name)
```
We need to escape this

aviv86 replied:
done

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesDetails.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public class TimeSeriesDetails
+    {
+        public string Id;
```
convert to properties. It is easier to inherit and override if necessary in the future.

aviv86 replied:
done

ppekrol reviewed (approved):
We need to rebase it

aviv86 committed:
RavenDB-14882 : minor fixes after rebase

aviv86 committed:
RavenDB-14882 : minor changes in TimeSeriesHandler :
- In ReadRanges endpoint fromList/toList should be required and cannot contain null/empty strings
- In Read endpoint make from/to non-required and default to DateTime.Min and DateTime.Max (studio uses that)

aviv86 committed:
RavenDB-14882 : address PR comments

aviv86 committed:
RavenDB-14882 : remove redundant GetMultipleTimeSeriesOperation overload with a single range (use GetTimeSeriesOperation instead)

aviv86 commented:
Rebased

karmeli87 commented on src/Raven.Server/Documents/Handlers/TimeSeriesHandler.cs:
```diff
@@ -145,33 +145,74 @@ public async Task Read()
         }
 
-        private static Dictionary<string, List<TimeSeriesRangeResult>> GetTimeSeriesRangeResults(DocumentsOperationContext context, string documentId, StringValues names, StringValues fromList, StringValues toList, int start, int pageSize)
+        [RavenAction("/databases/*/timeseries", "GET", AuthorizationStatus.ValidUser)]
```
I think this requires studio change as well
@ml054

ml054 replied:
oh yes - now I can remove those magic constants!
